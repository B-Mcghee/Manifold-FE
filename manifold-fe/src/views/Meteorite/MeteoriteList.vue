<template>
  <v-data-iterator
    :items="meteorites"
    :items-per-page.sync="itemsPerPage"
    :page.sync="page"
    :search="search"
    :sort-by="sortBy.toLowerCase()"
    :sort-desc="sortDesc"
    hide-default-footer
    class="relative"
  >
    <!-- <template v-slot:header></template> -->
    <template v-slot:header>
      <v-toolbar dark color="#1b579e" class="mb-1 fixed" height="50px">
        <v-text-field
          v-model="search"
          clearable
          flat
          solo-inverted
          hide-details
          prepend-inner-icon="mdi-magnify"
          label="Search"
          class="mr-3 fixed"
          fixed
        >
        </v-text-field>
        <template v-if="$vuetify.breakpoint.smAndUp">
          <v-select
            v-model="sortBy"
            flat
            solo-inverted
            hide-details
            :items="keys"
            prepend-inner-icon="mdi-magnify"
            label="Sort by"
          >
          </v-select>
          <v-spacer></v-spacer>
          <v-btn-toggle v-model="sortDesc" mandatory>
            <v-btn large depressed color="#1b579e" :value="true">
              <v-icon>mdi-arrow-up</v-icon>
            </v-btn>
            <v-btn large depressed color="#1b579e" :value="false">
              <v-icon>mdi-arrow-down</v-icon></v-btn
            >
          </v-btn-toggle>
        </template>
      </v-toolbar>
    </template>

    <template v-slot:default="props">
      <div class="overflow">
        <v-row>
          <v-col
            align-center
            v-for="item in props.items"
            :key="item.id"
            cols="12"
            sm="6"
            md="3"
            lg="3"
          >
            <meteorite-list-item :meteorite="item"></meteorite-list-item>
          </v-col>
        </v-row>
      </div>
    </template>

    <template v-slot:footer>
      <v-row
        class="mt-2 footer blue--text text--darken-2 blue-grey lighten-5"
        align="center"
        justify="center"
      >
        <span class="">Items per page</span>
        <v-menu offset-y v-model="showCaret">
          <template v-slot:activator="{ on, attrs }">
            <v-btn
              dark
              text
              color="primary"
              class="ml-4"
              v-bind="attrs"
              v-on="on"
            >
              {{ itemsPerPage }}
              <v-icon>mdi-chevron-down</v-icon>
            </v-btn>
          </template>
          <v-list>
            <v-list-item
              v-for="(number, index) in itemsPerPageArray"
              :key="index"
              @click="updateItemsPerPage(number)"
            >
              <v-list-item-title>{{ number }}</v-list-item-title>
            </v-list-item>
          </v-list>
        </v-menu>

        <v-spacer></v-spacer>

        <span
          class="mr-3
            grey--text"
        >
          Page {{ page }} of {{ numberOfPages }}
        </span>
        <v-btn
          fab
          dark
          color="blue darken-3"
          class="mr-3"
          height="2em"
          width="2em"
          @click="formerPage"
        >
          <v-icon>mdi-chevron-left</v-icon>
        </v-btn>
        <v-btn
          fab
          dark
          color="blue darken-3"
          class=" mr-2"
          height="2em"
          width="2em"
          @click="nextPage"
        >
          <v-icon>mdi-chevron-right</v-icon>
        </v-btn>
      </v-row>
    </template>
  </v-data-iterator>
</template>

<script>
// import MeteoriteListItem from "../components/MeteoriteListItem.vue";
import MeteoriteListItem from "./MeteoriteListItem.vue";
export default {
  components: { MeteoriteListItem },
  data() {
    return {
      showCaret: false,
      softDesc: false,
      meteors: null,
      meteorClass: null,
      search1: "",
      itemsPerPageArray: [25, 50, 100],
      search: "",
      filter: {},
      sortDesc: false,
      page: 1,
      itemsPerPage: 25,
      sortBy: "name",
      keys: [
        "id",
        "name",
        "nametype",
        "recclass",
        "mass",
        "fall",
        "year",
        "reclat",
        "reclong",
        "geolocation",
      ],
    };
  },
  // components: { MeteoriteListItem },

  created() {
    this.meteors = JSON.parse(JSON.stringify(this.meteorites));
  },
  computed: {
    meteorites: function() {
      const array = this.$store.getters.initialMeteorites.map(
        (meteorite) => meteorite
      );

      return array.slice(0, 5);
    },
    numberOfPages() {
      return Math.ceil(this.meteorites.length / this.itemsPerPage);
    },
    filteredKeys() {
      return this.keys.filter((key) => key !== "Name");
    },
    filteredMeteors() {
      if (this.search1.length < 1)
        return this.$store.getters.initialMeteorites.map(
          (meteorite) => meteorite
        );

      return this.meteorites.filter((meteor) =>
        meteor.name.toLowerCase().includes(this.search1.toLowerCase())
      );
    },
  },
  methods: {
    nextPage() {
      if (this.page + 1 <= this.numberOfPages) this.page += 1;
    },
    formerPage() {
      if (this.page - 1 >= 1) this.page -= 1;
    },
    updateItemsPerPage(number) {
      this.itemsPerPage = number;
    },
    resort(recclass) {
      console.log("resort");
      this.meteors = JSON.parse(JSON.stringify(this.meteorites));
      this.meteorClass = this.meteors.filter(
        (meteor) => meteor.recclass == recclass
      );
    },
  },
};
</script>

<style>
.card {
  border: 3px solid red;
}
fixed {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.overflow {
  height: 65vh;
  overflow: scroll;
}
.footer {
  padding: 4px 15px;
}
@media screen and (max-width: 1024px) {
  .overflow {
    height: 62vh;
    overflow: scroll;
  }
}
@media screen and (max-width: 1800px) {
  .overflow {
    height: 70vh;
    overflow: scroll;
  }
}
</style>
