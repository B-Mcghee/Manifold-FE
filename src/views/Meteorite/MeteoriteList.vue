<template>
  <div>
    <div
      class="  d-flex justify-center text-h5  pb-4 font-weight-bold primary--text text-lg-h4"
    >
      <h1 v-if="!year">{{ title }}</h1>
      <h1 v-else>{{ title }}</h1>
    </div>

    <meteorite-loader :meteorites="loadMeteorites"></meteorite-loader>
  </div>
</template>

<script>
import { mapGetters } from "vuex";
import MeteoriteLoader from "@/components/MeteoriteLoader.vue";
export default {
  props: {
    category: {
      type: Object
    }
  },
  components: { MeteoriteLoader },
  data() {
    return {
      year: false,
      title: "Meteorites"
    };
  },
  methods: {
    allMeteorites() {
      return this.initialMeteorites;
    },
    favoriteList() {
      const array = [];
      this.favorites.forEach(m => {
        this.initialMeteorites.filter(meteorite => {
          if (m == meteorite.id) {
            array.push(meteorite);
          }
        });
      });
      return array;
    },
    comparisonList() {
      const array = [];
      this.meteoriteComparison.forEach(m => {
        this.initialMeteorites.filter(meteorite => {
          if (m == meteorite.id) {
            array.push(meteorite);
          }
        });
      });
      return array;
    },
    propertyList(value) {
      value === "year"
        ? this.alteredTitle(`${this.category.value.substring(0, 4)} /`)
        : this.alteredTitle(`${this.category.value} /`);

      const array = this.initialMeteorites.filter(
        meteorite =>
          meteorite[`${this.category.property}`] == this.category.value
      );
      return array;
    },
    alteredTitle(value) {
      this.title = value + " Meteorites";
    }
  },

  computed: {
    ...mapGetters(["favorites", "initialMeteorites", "meteoriteComparison"]),
    loadMeteorites: function() {
      if (this.category.category) {
        switch (this.category.category) {
          case "Compare":
            this.alteredTitle("Compare");
            return this.comparisonList();
          case "Favorites":
            this.alteredTitle("Favorite");
            return this.favoriteList();
          default:
            break;
        }
      } else if (this.category.property) {
        return this.propertyList(this.category.property);
      }
      this.alteredTitle("All");
      return this.allMeteorites();
    }
  },
  filters: {}
};
</script>

<style></style>
