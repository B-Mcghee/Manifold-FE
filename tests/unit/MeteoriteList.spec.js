import { createLocalVue, shallowMount, mount } from "@vue/test-utils";
import Vuetify from "vuetify";
import "regenerator-runtime";

import Vuex from 'vuex'


import MeteoriteList from '@/views/Meteorite/MeteoriteList.vue'


const localVue = createLocalVue();
localVue.use(Vuex);


describe("MeteoriteList.vue",  () => {

  let wrapper;
  let vuetify;
  let store;
let getters;
let state;
let router;
let category;

  beforeEach(() => {
    category = {property:'recclass', value: 'L5'}
      getters = {
            favorites: (state) => state.favoriteMeteorites,
            initialMeteorites: (state) => state.meteorites,
         meteoriteComparison: (state) => state.meteoriteComparison
      }
      state = {
          meteorites:[
            {
                name: "Aachen",
                id: "1",
                nametype: "Valid",
                recclass: "L5",
                mass: "21",
                fall: "Fell",
                year: "1880-01-01T00:00:00.000",
                reclat: "50.775000",
                reclong: "6.083330",
                geolocation: {
                  type: "Point",
                  coordinates: [6.08333, 50.775]
                }
              },
              {
                name: "Abee",
                id: "6",
                nametype: "Valid",
                recclass: "EH4",
                mass: "107000",
                fall: "Fell",
                year: "1952-01-01T00:00:00.000",
                reclat: "54.216670",
                reclong: "-113.000000",
                geolocation: {
                  Type: "Point",
                  coordinates: [
                    -113,
                    54.21667
                  ]
                }
              }
          ],
          favoriteMeteorites:[1],
          meteoriteComparison:[2,3]
      }
    store = new Vuex.Store({
        getters,
        state
    })
    vuetify = new Vuetify();
  });
  it("renders meteorite",  () => {


    wrapper = shallowMount(MeteoriteList, {
     localVue,
     vuetify,
       store,
     data() {
       return {
         favorite: false
       };
     },
     propsData: {
       category
     },
         stubs:['router-link', 'router-view']
   });
   expect(wrapper.html()).toMatchSnapshot();
 });

it("renders meteorites from state, and getters",  () => {


    wrapper = shallowMount(MeteoriteList, {
     localVue,
     vuetify,
       store,

     data() {
       return {
         favorite: false
       };
     },
     propsData: {
       category
     },
         stubs:['router-link', 'router-view']
   });
   expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.vm.allMeteorites()).toBe(store.state.meteorites)




 });
 it("testing comparison list method to match up with state",  () => {


    wrapper = shallowMount(MeteoriteList, {
     localVue,
     vuetify,
       store,

     data() {
       return {
         favorite: false
       };
     },
     propsData: {
       category
     },
         stubs:['router-link', 'router-view']
   });
const compare = [];
store.getters.meteoriteComparison.forEach((m) => {
  store.getters.initialMeteorites.filter((meteorite) => {
    if (m == meteorite.id) {
      compare.push(meteorite);
    }
  });
})
expect(wrapper.vm.comparisonList()).toStrictEqual(compare)
})
it("testing comparisonList method",  () => {


    wrapper = shallowMount(MeteoriteList, {
     localVue,
     vuetify,
       store,

     data() {
       return {
         favorite: false
       };
     },
     propsData: {
       category
     },
         stubs:['router-link', 'router-view']
   });
const array = []; 
store.getters.favorites.forEach((m) => {
    store.getters.initialMeteorites.filter((meteorite) => {
      if (m == meteorite.id) {
        array.push(meteorite);
      }
    })
})

expect(wrapper.vm.favoriteList()).toStrictEqual(array)
});
})