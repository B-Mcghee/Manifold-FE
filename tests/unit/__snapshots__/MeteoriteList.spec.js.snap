// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeteoriteList.vue renders meteorite 1`] = `
<div>
  <div class="  d-flex justify-center text-h5  pb-4 font-weight-bold primary--text text-lg-h4">
    <h1>Meteorites</h1>
  </div>
  <meteorite-loader-stub meteorites="[object Object]"></meteorite-loader-stub>
</div>
`;

exports[`MeteoriteList.vue renders meteorites from state, and getters 1`] = `
<div>
  <div class="  d-flex justify-center text-h5  pb-4 font-weight-bold primary--text text-lg-h4">
    <h1>Meteorites</h1>
  </div>
  <meteorite-loader-stub meteorites="[object Object]"></meteorite-loader-stub>
</div>
`;
